SDWebImage/SDWebImageDownloader.m

```diff
@@ -307,6 +307,15 @@ - (nullable SDWebImageDownloadToken *)downloadImageWithURL:(nullable NSURL *)url
         // `addOperation:` does not synchronously execute the `operation.completionBlock` so this will not cause deadlock.
         [self.downloadQueue addOperation:operation];
     }
+    else if (!operation.isExecuting) {
```

Comment:
...I don't understand what's the reason, to put these code inside the synchonized lock. Does this modify the operation queue and cause thread-safe issue ?

Reply:
@dreampiggy Because we can reuse the state of operation, operation has 5 states, nil、ready not start、executing、finished、cancelled, we already check three states at [here](https://github.com/SDWebImage/SDWebImage/pull/2575/files#diff-863e5ca8f79e3f53ae0acb6bc11c2b56L293) or see #2574 , and for most situations, operation would hit these three states, so we only need to add `isExecuting` check, if it not executing, means it just ready to run, we can assign the priority. 

For your `PR` #2562 , you always do the check for every operation, it's redundancy, because for new created operation, we don't need to do it.

Reply:
My consideration is performance, not whether the code is clean or not. It’s the PR’s author’s responsibility to make better code.

If you mean “I don’t want to check that condition twice, it make my code ugly”, you can write a bool variable to check. It’s not hard and does not impact current performance

Reply:
Where is the performance penalty? only because I put it in a lock? if you worry about it, profile it to show the statistics please.

PS. Besides code place, your `PR` also has bug, see [comment](https://github.com/SDWebImage/SDWebImage/pull/2575#discussion_r246410267)

Reply:
Emm it may very hard to talk, I say it has 5 states, I don't say at some code line, it would not be nil.

If you should put only thread-safe operation in synchronization, the below code only `createDownloaderOperationWithUrl` and `[self.URLOperations setObject:operation forKey:url];` should put in synchronize lock. other line need put outside. That dose not need synchronized.

https://github.com/SDWebImage/SDWebImage/blob/8a743950e26af56f13c7662b0b6a68f19385c136/SDWebImage/SDWebImageDownloader.m#L294-L308

I don't understand why this is unrelated code, it's the initialize operation.

Reply:
> it's the initialize operation.

This is not initialize operation. 😕 If this is a initialize operation, why not place all of this code into the `createDownloaderOperationWithUrl:options:` ?

This is a logic applied for `something after we quired one valid operation`.

Reply:
@dreampiggy `createDownloaderOperationWithUrl:options:` has the same code, check it.